Zend/zend_compile.c

```diff
@@ -8795,6 +8801,15 @@ void zend_compile_instanceof(znode *result, zend_ast *ast) /* {{{ */
 	zend_compile_class_ref(&class_node, class_ast,
 		ZEND_FETCH_CLASS_NO_AUTOLOAD | ZEND_FETCH_CLASS_EXCEPTION);
 
+	if (class_node.op_type == IS_CONST) {
+		zend_uchar t = zend_lookup_builtin_type_by_name(Z_STR(class_node.u.constant));
+
+		if (t == IS_NULL || t == _IS_BOOL || t == IS_LONG || t == IS_STRING || t == IS_DOUBLE || t == IS_OBJECT) {
+			zend_compile_func_typecheck_impl(result, obj_ast, t);
+			return;
+		}
+	}
+
 	opline = zend_emit_op_tmp(result, ZEND_INSTANCEOF, &obj_node, NULL);
```

Comment:
ZEND_INSTANCEOF would have to be adjusted to make sure that you get the same behavior with a dynamic type name.